client/consensus/slots/src/lib.rs

```diff
@@ -298,20 +284,25 @@ pub trait SimpleSlotWorker<B: BlockT> {
 
 		let logs = self.pre_digest_data(slot, &claim);
 
-		// deadline our production to approx. the end of the slot
+		// deadline our production to 98% of the total time left for proposing. As we deadline
+		// the proposing below to the same total time left, the 2% margin should be enough for
+		// the result to be returned.
 		let proposing = awaiting_proposer.and_then(move |proposer| proposer.propose(
 			slot_info.inherent_data,
 			sp_runtime::generic::Digest {
 				logs,
 			},
-			slot_remaining_duration,
+			proposing_remaining_duration.mul_f32(0.98),
```

Comment:
Is there a risk that the 2% ends up being too small when the remaining duration becomes very small? As opposed to use a % of the proposing time or slot time which should be a pretty stable duration?

Reply:
That is a good question. For parachains this will currently be 500ms for proposing, so leaving us with 10ms buffer. I would say that this is enough buffer, especially as the select is biased to the `proposing`, meaning this will be polled first. 
Do you have some better idea? 

I think we should use this for now and adapt if someone complains. 

WDYT @andresilva ?

Reply:
For the default case, but with a lenience factor applied, the timeout was always > than the remaining time of the slot. (Before if there was a runtime upgrade that maybe took 7 seconds and a proposing time of 8 seconds after applying the lenience factor, we would have build the block and ended the proposing. Now, it could theoretically happen that we still add one or 2 transactions :grimacing: as we propagate the real proposing time to the block builder.)